### PR TITLE
docs(spanner): document `RetryPolicy` interface

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #9434

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12037)

### CODE REVIEW

File: google/cloud/spanner/retry_policy.h [hunk 1]
```diff
@@ -59,33 +59,267 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
...
-        spanner_internal::SafeTransactionRerun>;
+/// The base class for the Spanner library retry policies.
+class RetryPolicy : public google::cloud::RetryPolicy {
+ public:
+  /// Creates a new instance of the policy, reset to the initial state.
+  virtual std::unique_ptr<RetryPolicy> clone() const = 0;
+};
+
+/**
+ * A retry policy for `$connection_class_name$` based on counting errors.
```
Comment (devbww) [inline google/cloud/spanner/retry_policy.h:70]: `$connection_class_name$`?

File: google/cloud/spanner/retry_policy.h [hunk 2]
```diff
@@ -59,33 +59,267 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
...
+  bool IsExhausted() const override { return impl_.IsExhausted(); }
+  bool IsPermanentFailure(Status const& s) const override {
+    return impl_.IsPermanentFailure(s);
+  }
+  std::unique_ptr<TransactionRerunPolicy> clone() const override {
+    return std::make_unique<LimitedTimeTransactionRerunPolicy>(
+        impl_.maximum_duration());
+  }
+
+  // Not very useful, but needed for backwards compatibility.
```
Comment (devbww) [inline google/cloud/spanner/retry_policy.h:315]: Nit: Use the same "This is provided only for backwards compatibility" text as the other ones?

**Review by devbww — APPROVED** [review]

File: google/cloud/spanner/retry_policy.h (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/retry_policy.h]: ooops, done.

File: google/cloud/spanner/retry_policy.h (hunk 2 above)
Comment (coryan) [reply google/cloud/spanner/retry_policy.h]: Fixed, and added one missing typedef.

**Pull request merged**